### PR TITLE
encoding/wkb: new type Geom

### PR DESCRIPTION
**Pull request by tbruyelle**
When the shape of a geometry isn't defined (for example when you use the
GEOMETRY postgis type), you can't use the scanners defined in
encoding/wkb, you have to iterate over the results and invoke
wkb.Marshal on the column.

This change introduces a new type wkb.Geom which can scan any kind of
geometry shape.

I've added NOTEs about details I would like to discuss, please review
them.

I've also added new examples to illustrate the usage, tell me if the
example names are OK for you.

### CODE REVIEW

File: encoding/wkb/sql.go [hunk 1]
```diff
@@ -18,6 +18,16 @@ func (e ErrExpectedByteSlice) Error() string {
 	return fmt.Sprintf("wkb: want []byte, got %T", e.Value)
 }
 
+// A Geom is a WKB-ecoded Geometry that implements the sql.Scanner and
+// driver.Value interfaces.
+// It can be used when the geometry shape is not defined.
+type Geom struct {
+	geom.T
+	// NOTE(tb) I've repeated the opts field like other types, but I'm not sure
+	// about its usage.
```
Comment (twpayne) [inline encoding/wkb/sql.go:27]: These options control how the geometry is encoded, specifically how empty points are handled.

File: encoding/wkb/sql.go [hunk 2]
```diff
@@ -67,6 +77,32 @@ type GeometryCollection struct {
 	opts []wkbcommon.WKBOption
 }
 
+// Scan scans from a []byte.
+func (g *Geom) Scan(src interface{}) error {
+	b, ok := src.([]byte)
+	if !ok {
+		return ErrExpectedByteSlice{Value: src}
+	}
+	// NOTE(tb) other Scanners do not check the len of b, is it really useful ?
```
Comment (twpayne) [inline encoding/wkb/sql.go:86]: I need to double check this. Specifically, this relates to how `NULL` `GEOMETRY`s and `GEOGRAPHIES` should be handled. `database/sql` has a selection of `Null*` types (e.g. `NullBool`, `NullString` etc.) and it's not yet clear to me whether we should include `NULL` handling in `wkb.Geom` by default or if we should add a separate `wkb.NullGeom`.

tl;dr please leave this in for the moment.

File: encoding/wkb/sql.go [hunk 3]
```diff
@@ -67,6 +77,32 @@ type GeometryCollection struct {
...
+	return err
+}
+
+// Value returns the WKB encoding of g.
+func (g *Geom) Value() (driver.Value, error) {
+	return value(g.T)
+}
+
+// NOTE(tb) you asked for this method but since accessing the geom.T is trivial
+// as writing g.T, I wonder if it worth it.
```
Comment (twpayne) [inline encoding/wkb/sql.go:101]: I agree it's trivial, but I think it will look prettier in code as it better indicates what is returned compared to just `.T` alone.

File: encoding/wkb/sql_example_test.go [hunk 4]
```diff
@@ -74,3 +74,98 @@ func Example_value() {
...
+	if err != nil {
+		log.Fatal(err)
+	}
+	for rows.Next() {
+		var s Shape
+		err := rows.Scan(&s.Name, &s.Geom)
+		if err != nil {
+			log.Fatal(err)
+		}
+		fmt.Printf("%s: %v\n", s.Name, s.Geom.FlatCoords())
```
Comment (twpayne) [inline encoding/wkb/sql_example_test.go:108]: nit: please print `s.Geom.Coords()` instead as it makes the differences between the types clearer.

File: encoding/wkb/sql_example_test.go [hunk 5]
```diff
@@ -74,3 +74,98 @@ func Example_value() {
...
+		Geom wkb.Geom
+	}
+
+	db, mock, err := sqlmock.New()
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer db.Close()
+
+	mock.ExpectQuery(`SELECT name, ST_AsBinary\(geom\) FROM shapes`).
```
Comment (twpayne) [inline encoding/wkb/sql_example_test.go:90]: I don't think `(` and `)` need to be escaped.

File: encoding/wkb/sql_example_test.go [hunk 6]
```diff
@@ -74,3 +74,98 @@ func Example_value() {
...
+		fmt.Printf("%s: %v\n", s.Name, s.Geom.FlatCoords())
+	}
+
+	// Output:
+	// Point: [1 2]
+	// LineString: [1 2 3 4]
+	// Polygon: [1 2 3 4 5 6 1 2]
+}
+
+func Example_value_different_shapes() {
```
Comment (twpayne) [inline encoding/wkb/sql_example_test.go:117]: Same comments removing `\`s from queries and printing `Coords` instead of `FlatCoords` apply.

File: encoding/wkb/sql_example_test.go (hunk 4 above)
Comment (tbruyelle) [reply encoding/wkb/sql_example_test.go:108]: but geom.T has no `Coords()` method :)

File: encoding/wkb/sql_example_test.go (hunk 5 above)
Comment (tbruyelle) [reply encoding/wkb/sql_example_test.go:90]: If I remove them the test fail, I think it's related to an internal regexp usage in `go-sqlmock`. Moreover the other existing examples use the same escapes.

File: encoding/wkb/sql.go (hunk 1 above)
Comment (tbruyelle) [reply encoding/wkb/sql.go]: ok thanks, I removed the note 
92b5cec

File: encoding/wkb/sql.go (hunk 3 above)
Comment (tbruyelle) [reply encoding/wkb/sql.go]: Fine, I replaced the note with a godoc

92b5cec

File: encoding/wkb/sql_example_test.go (hunk 5 above)
Comment (twpayne) [reply encoding/wkb/sql_example_test.go:90]: OK, thanks.

File: encoding/wkb/sql_example_test.go (hunk 4 above)
Comment (twpayne) [reply encoding/wkb/sql_example_test.go:108]: Fair point!

**Comment by twpayne** [comment]
The PR looks good, there's just a bit of missing punctuation in a godoc comment. If you don't want to fix this I can fix it.

**Comment by tbruyelle** [comment]
> The PR looks good, there's just a bit of missing punctuation in a godoc comment. If you don't want to fix this I can fix it.

fixed :)

**Comment by twpayne** [comment]
Thanks for this really nice contribution :)
